test/object-comprehensions.civet

```diff
@@ -0,0 +1,266 @@
+    scoping
+    ---
+    {
+      for results of [1]
+        [results]:  2 * x
+    }
+    ---
+    ({
+      ...(()=>{const results1={};for (const results of [1]) {
+        Object.assign(results1,({[results]:  2 * x}))
+      }return results1})()
+    })
+  """
+
+  testCase """
+    loop body with additional statements: single prop
+    ---
+    {
+      for x of [1]
+        foo bar
+        [x]:  2 * x
+    }
+    ---
+    ({
+      ...(()=>{const results={};for (const x of [1]) {
+        foo(bar)
+        Object.assign(results,({[x]:  2 * x}))
+      }return results})()
+    })
+  """
+
+  testCase """
+    loop body with additional statements: multi prop style 1
+    ---
+    {
+      for x of [1]
+        foo bar
+        {
+          [x]:  2 * x
+          [2*x]: 4 * x
+        }
+    }
+    ---
+    ({
+      ...(()=>{const results={};for (const x of [1]) {
+        foo(bar)
+        Object.assign(results,({
+          [x]:  2 * x,
+          [2*x]: 4 * x
+        }))
+      }return results})()
+    })
+  """
+
+  testCase """
+    loop body with additional statements: multi prop style 2
+    ---
+    {
+      for x of [1]
+        foo bar
+        [x]:  2 * x
+        [2*x]: 4 * x
+    }
+    ---
+    ({
+      ...(()=>{const results={};for (const x of [1]) {
+        foo(bar)
+        Object.assign(results,({[x]:  2 * x,
+        [2*x]: 4 * x}))
+      }return results})()
+    })
+  """
+
+  testCase """
+    on first line
+    ---
+    {for x of [1]
+      [x]:  2 * x
+      [2*x]: 4 * x }
+    ---
+    ({...(()=>{const results={};for (const x of [1]) {
+      Object.assign(results,({[x]:  2 * x,
+      [2*x]: 4 * x}))
+    }return results})() })
+  """
+
+  testCase """
+    on first line with other props
+    ---
+    { a: "a prop", for x of [1]
+      [x]:  2 * x
+      [2*x]: 4 * x }
+    ---
+    ({ a: "a prop", ...(()=>{const results={};for (const x of [1]) {
+      Object.assign(results,({[x]:  2 * x,
+      [2*x]: 4 * x}))
+    }return results})() })
+  """
+
+  testCase """
+    single line
+    ---
+    o := {for x of [1] [x]: 2 * x }
+    ---
+    const o = {...(()=>{const results={};for (const x of [1]({[x]: 2 * x})) {Object.assign(results,x)}return results})() }
+  """
+
+  testCase """
+
+    non-comprehension loop
+    ---
+    o := {...for x of [1] [x]: 2 * x }
+    ---
+    const o = {...(()=>{const results=[];for (const x of [1]({[x]: 2 * x})) {results.push(x)}return results})() }
+  """
+
+  testCase """
+    while object comprehension
+    ---
+    o := { while(predicate(x)) [x]: f(x)}
+    ---
+    const o = { ...(()=>{const results={};while(predicate(x)) Object.assign(results,({[x]: f(x)}));return results})()}
+  """
+
+  testCase """
+    non-comprehension while loop
+    ---
+    o := {...while(predicate(x)) [x]: f(x)}
+    ---
+    const o = {...(()=>{const results=[];while(predicate(x)) results.push(({[x]: f(x)}));return results})()}
+  """
+
+  testCase """
+    do while object comprehension
```

Comment:
@edemaine I discovered one fundamental design issue here.

While this one-liner do while loop works fine, I cannot support a multi-line braceless do-while loop.

So for instance, this works:
```
o := {
  do {
    [x]: f(x)
  } while(predicate(x)) 
}
```

But this does not work:

```
o := {
  do
    [x]: f(x)
  while(predicate(x)) 
}
```

When there is nothing after `do`  except a newline, it's interpreted as a `NamedProperty`. This does not affect for and while because they have some tokens before a newline.

I can prioritize my rule by putting it above all the other rules in `PropertyDefinition`, and surprisingly that breaks only one test!

But that test is https://github.com/DanielXMoore/Civet/blob/b084f871269562be1e3ef12f8fe8f114577f5a44/test/object.civet#L1409 , which is about Civet's feature that there are no reserved words. 

If I prioritize loops, then "do" becomes a reserved word. If I don't, I can't support multi-line do.

Reply:
Good catch! Yes, I think you should move your rule up for higher priority, and remove the looping reserved words from the test.